org/mozilla/jss/ssl/common.c

```diff
@@ -899,3 +900,201 @@ JSS_SSL_processExceptions(JNIEnv *env, PRFilePrivate *priv)
+
+    for (node = CERT_LIST_HEAD(certList);
+       !CERT_LIST_END(node, certList);
+       node = CERT_LIST_NEXT(node)) {
+       
+        /* try to find the root */
+       if( node->cert && node->cert->isRoot ) {
+          root = CERT_DupCertificate(node->cert) ;
+       } 
+    }
+
+finish:
+  
+    CERT_DestroyCertList (certList); 
+    return root; 
+}
+
+SECStatus JSSL_verifyCertPKIX(CERTCertificate *cert,
+      SECCertificateUsage certUsage,secuPWData *pwdata, int ocspPolicy,
+      CERTVerifyLog *log, SECCertificateUsage *usage) 
+{
+
+    /* put the first set of possible flags internally here first */
+    /* later there could be a more complete list to choose from */
+    /* support our hard core fetch aia ocsp policy for now */
+
+    static PRUint64 ocsp_Enabled_Hard_Policy_LeafFlags[2] = {
+        /* crl */
+        0,
+        /* ocsp */
+        CERT_REV_M_TEST_USING_THIS_METHOD |
+        CERT_REV_M_FAIL_ON_MISSING_FRESH_INFO
+    };
+
+    static PRUint64 ocsp_Enabled_Hard_Policy_ChainFlags[2] = {
+        /* crl */
+        0,
+        /* ocsp */
+        CERT_REV_M_TEST_USING_THIS_METHOD |
+        CERT_REV_M_FAIL_ON_MISSING_FRESH_INFO
+    };
+
+    static CERTRevocationMethodIndex
+        ocsp_Enabled_Hard_Policy_Method_Preference = {
+            cert_revocation_method_ocsp
+        };
+
+    static CERTRevocationFlags ocsp_Enabled_Hard_Policy = {
+    { /* leafTests */
+      2,
+      ocsp_Enabled_Hard_Policy_LeafFlags,
+      1,
+      &ocsp_Enabled_Hard_Policy_Method_Preference,
+      0 },
+    { /* chainTests */
+      2,
+      ocsp_Enabled_Hard_Policy_ChainFlags,
+      1,
+      &ocsp_Enabled_Hard_Policy_Method_Preference,
+      0 }
+    };
+
+    /* for future expansion */
+
+    CERTValOutParam cvout[20] = {0};
+    CERTValInParam cvin[20] = {0};
+
+    int inParamIndex = 0;
+    int outParamIndex = 0;
+    CERTRevocationFlags *rev = NULL;
+
+    CERTCertList *trustedCertList = NULL;
+
+    PRBool fetchCerts = PR_FALSE;
+
+    SECStatus res =  SECFailure;
+    if(cert == NULL) {
+        goto finish;
+    }
+
+    if(ocspPolicy != OCSP_LEAF_AND_CHAIN_POLICY) {
+        goto finish;
+    }
+
+    /* Force the strict ocsp network check on chain
+       and leaf.
+    */
+
+    if(ocspPolicy == OCSP_LEAF_AND_CHAIN_POLICY) {
+        fetchCerts = PR_TRUE;   
+        rev = &ocsp_Enabled_Hard_Policy;
+    }  else {
+        /* for now */
+        goto finish;
+    }
+  
+    /* fetch aia over net */
+ 
+    cvin[inParamIndex].type = cert_pi_useAIACertFetch;
+    cvin[inParamIndex].value.scalar.b = fetchCerts;
+    inParamIndex++; 
+
+    /* time */
+
+    cvin[inParamIndex].type = cert_pi_date;
+    cvin[inParamIndex].value.scalar.time = PR_Now();
+    inParamIndex++;
+
+    /* flags */
+
+    cvin[inParamIndex].type = cert_pi_revocationFlags;
+    cvin[inParamIndex].value.pointer.revocation = rev;
+    inParamIndex++;
+
+
+    /* establish trust anchor */
+
+    CERTCertificate *root = getRoot(cert,certUsage);
```

Comment:
If `root` is NULL, is there any meaningful work we can do? Or should we just `goto finish`?

Reply:
If root is NULL, the system will just go and pick a "trusted root" for us. It may end up being the intermediate cert, which is not what we want , but it will go ahead and check probably the leaf, with is still useful. So I think this is reasonable.

Reply:
Ok, thanks for the explanation.